common/src/primitives/merkle/tree/mod.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod tree_size;
+
+use std::num::NonZeroUsize;
+
+use crypto::hash::StreamHasher;
+
+use crate::primitives::{
+    id::{default_hash, DefaultHashAlgoStream},
+    H256,
+};
+
+use self::tree_size::TreeSize;
+
+use super::{pos::NodePosition, MerkleTreeAccessError, MerkleTreeFormError};
+
+pub enum AdjacentLeavesIndices {
+    Alone(usize),
+    Together(usize, usize),
+}
+
+/// Merkle tree in the form of a vector, where the bottom leaves first, from left to right, and the root is
+/// the last element.
+/// Definitions:
+/// - Leaf: A node does not have children.
+/// - Node: any element of the tree.
+/// - Absolute index: The data structure that represents the tree internally is a Vec.
+///   Leaves start from 0 and from the left and end at the right, then next, higher levels. The last element is the root.
+///   The absolute index is the index to find a specific node in this tree, regardless of where it lies.
+/// - Position: The "coordinates" to find a node, given as level (leaves are at level 0, root is highest level),
+///   and index (we count from left to right).
+/// - Root: The root of the tree; in merkle-tree's case, it's the node that's created by hashing all the elements underneath.
+/// - Padding: Extra elements we add to the tree to make the number of leaves a power of 2. This has to match some security specs.
+///
+/// Given that this is strictly a filled-up binary tree, the number of leaves is always a power of 2, and the total number of
+/// nodes is always 2 * leaves - 1. These are invariants that are always held through type-level checks.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct MerkleTree {
+    tree: Vec<H256>,
+}
+
+impl MerkleTree {
```

Comment:
I don't know... I'm hesitating to do this because reasoning about their security will be harder.

Reply:
I am willing to accept this but I am not sure I see the issue. In what sense will reasoning about security become harder? Is the concern computational complexity or is it something else? Would appreciate some examples / pointers.

Reply:
Consider the padding. The padding will require creating a Vec by scanning the Vec itself. This is a mess in iterators. I'm not super-comfortable with that to the point that I can be sure I didn't do something wrong. So, this being simple for a for-loop guarantees that I won't do a mistake. It's much easier for others to read as well.

The `from_leaves` function is easy to make iterators. I'm happy to conceded that much. But without doing that for padding... not a big win.

Having said that, any attempt to generalize the padding process will ruin anything we do here. I still have a plan to do generics for later. That will involve a separate type for padding. Doing this with iterators will just make this harder if we consider doing that, because the operations that we do with iterators will be specialized for this padding. It's much easier to deal with (and think of) a `Vec<H256>` than an iterator when future plans are not clear.

If you insist I'll give it a shot. But I'm not really comfortable with this.

Reply:
> Consider the padding. The padding will require creating a Vec by scanning the Vec itself. This is a mess in iterators.

Regarding how padding could be done with iterators, something like the following should work:
```rust
let padding = std::iter::successors(elements.last(), |last_pad| default_hash(last_pad));
let tree_leaves = elements.iter().chain(padding).take(pow2_size);
```

I am not sure what are your plans with generics so cannot comment on that.

Reply:
I don't think this will work. You can't access `elements.last()` without iterating through the iterator. This will create a circular dependency. You can't chain without getting the padding, and you can't pad without getting the last element. So now we have to peek somehow to both pad and check whether the count of elements is > 0. So basically you'll be unrolling your iterator under the hood and converting it back to an iterator to chain it with the padding. I'm not sure how much of an improvement that is to a Vec.

Again, maybe I'm missing something or I'm misunderstanding some of the possible magic in iterators.

Reply:
Yea the suggestion is a bit more limited, assumes `elements` is still a slice. Just the way padding and leaves are stitched together can be turned into iterators, without the extra intermediate `Vec`s.

Reply:
So, I created a wrapping iterator. But unfortunately, there's still two allocations because of the circular dependency we talked about.

The ultimate solution will be to build the whole tree from an iterator.

Reply:
> So, I created a wrapping iterator. But unfortunately, there's still two allocations because of the circular dependency we talked about.
> 
> The ultimate solution will be to build the whole tree from an iterator.

Yea that s not very straightforward. This looks good though.